docs/sources.md

```diff
@@ -0,0 +1,268 @@
+`id`
+
+This is expected to be idempotent, a read-only activity that returns a unique cache key.
+
+The arguments to `CacheKey()` are:
+
+* `ctx context.Context`: A golang [context](https://pkg.go.dev/context#Context).
+* `g session.Group`: The [session.Group](https://pkg.go.dev/github.com/moby/buildkit@v0.10.1/session#Group) contains parameters to identify the specific buildkit client calling this build. These may be useful for things like credentials.
+* `index int`: The `CacheKey()` can be called multiple times for the same source item, each indicating that it is a separate subitem. The `index` indicates which iteration of call this is. Generally, if the `index > 0` then it is complete.
+
+The expected return values are:
+
+* `cacheKey string`: the unique cache ID; this might be the combination of a hash with the source identifier. For example, `docker-image` uses the hash of the image config, while `local` combines the path to the directory and hash of a json that includes the session and search parameters.
+* `imgDigest string`: the digest of the root of the image. For example, `docker-image` uses the hash of the root manifest or index, while `local` returns the hash of the constructed json of the session and search parameters.
+* `cacheOpts solver.CacheOpts`: The [CacheOpts](https://pkg.go.dev/github.com/moby/buildkit@v0.10.1/solver#CacheOpts) contain metadata that can be passed from one iteration of `CacheKey()` to another, enabling context to be passed from call to call without the `Source` needing to store information. This is useful for items like progress streams and remote reference information.
+* `cacheDone bool`: Indicates whether or not caching is complete.
+* `err error`: error, if any.
+
+#### `Snapshot`
+
+`Snapshot()` is what the solver calls to retrieve the content for the `id`. Normally, this gets run if it turns out there is no cached result. It does the work to
+actually run the operation and create the underlying cache reference in the buildkit worker.
+
+Note that it is not absolutely necessary to retrieve the entire content when `Snapshot()` is called. It is possible to do a "lazy retrieval". For example,
+`docker-image` can use lazy retrieval, and defer pulling layers until they really are needed. However, everything that is needed to retrieve the content gets set up
+by `Snapshot()`.
+
+For example, `docker-image` pulls the image, while `local` ???
+
+The arguments to `Snapshot()` are:
+
+* `ctx context.Context`: golang [context](https://pkg.go.dev/context#Context)
+* `g session.Group`: The [session.Group](https://pkg.go.dev/github.com/moby/buildkit@v0.10.1/session#Group) contains parameters to identify the specific buildkit client calling this build. These may be useful for things like credentials.
+
+The expected return values are:
+
+* `cache.ImmutableRef`: a [cache.ImmutableRef interface](../cache/refs.go#L48), which is used to read the actual content.
```

Comment:
This isn't a `io.Reader` or `TarReader`, it is an `ImmutableRef`. What does that mean? How does it use it to get the actual content in a useful way?

Reply:
OK, so it isn't the `SourceInstance` passing a stream of content that the solver will cache in its own cache, mount, etc.

Instead, it is that the _SourceInstance_ is expected to cache it locally, returning instead a struct that implements `ImmutableRef`, i.e. something that the solver can use to mount the content in the filesystem.

So then what does the whole `ImmutableRef` interface do? The interface is:

```go
type ImmutableRef interface {
	Ref
	Clone() ImmutableRef
	// Finalize commits the snapshot to the driver if it's not already.
	// This means the snapshot can no longer be mounted as mutable.
	Finalize(context.Context) error

	Extract(ctx context.Context, s session.Group) error // +progress
	GetRemotes(ctx context.Context, createIfNeeded bool, cfg config.RefConfig, all bool, s session.Group) ([]*solver.Remote, error)
	LayerChain() RefList
}
```

What is the `SourceInstance` expected to provide here? There's a lot more than "give me a mount point and I will mount it."